### ISSUE TITLE
TabNetRegressor.fit() returns None for history object despite successful training epochs

### ISSUE THREAD

**Issue by lucalazzaro** [dreamquark-ai/tabnet#579]
**Describe the bug**
When training a TabNetRegressor model, the .fit() method consistently returns None for the history_obj, rather than the expected History object. Despite this, the training process appears to run successfully, logging epochs, loss, and validation metrics, and early stopping is correctly triggered. No exceptions or errors are raised during the fit() execution itself.

**What is the current behavior?**
The TabNetRegressor.fit() method returns None. Consequently, there is no history attribute to access training progress programmatically.

**If the current behavior is a bug, please provide the steps to reproduce.**
Create a Python script named test_tabnet.py with the following content:

Python

import numpy as np
from pytorch_tabnet.tab_model import TabNetRegressor

print("Inizio test TabNet...")

# Generate some dummy data for regression
X = np.random.rand(100, 10).astype(np.float32) # 100 samples, 10 features
y = np.random.rand(100, 1).astype(np.float32) # 100 samples, 1 target

# Define model parameters
# Using minimal parameters for a quick test
model_params = {
    'n_d': 8,
    'n_a': 8,
    'n_steps': 3,
    'gamma': 1.3,
    'cat_emb_dim': 1,
    'n_independent': 2,
    'n_shared': 2,
    'lambda_sparse': 1e-3,
    'momentum': 0.02,
    'clip_value': 1,
    'optimizer_fn': None, # Default to Adam
    'optimizer_params': {'lr': 2e-2},
    'scheduler_fn': None, # No scheduler
    'scheduler_params': {},
    'mask_type': 'sparsemax',
    'verbose': 100,
    'seed': 42,
    'device_name': 'cpu' # Force CPU to avoid CUDA issues
}

# Initialize TabNetRegressor
regressor = TabNetRegressor(**model_params)

print("Avvio addestramento del modello TabNet...")

# Train the model
# Using a small number of epochs for quick testing
# Using X and y directly for simplicity, assuming they are already split
max_epochs = 2 # Small number of epochs to demonstrate the issue quickly
patience = 5 # Patience for early stopping
batch_size = 16

history_obj = regressor.fit(
    X_train=X,
    y_train=y,
    eval_set=[(X, y)], # Using training set as eval set for simplicity
    eval_name=['val_0'],
    max_epochs=max_epochs,
    patience=patience,
    batch_size=batch_size,
    virtual_batch_size=batch_size,
    num_workers=0, # Set to 0 for Windows compatibility
    drop_last=False
)

print("\n--- RISULTATO DEL TEST ---")
print(f"Tipo di history_obj: {type(history_obj)}")
print(f"Valore di history_obj: {history_obj}")
print(f"history_obj è None? {history_obj is None}")
if history_obj is None:
    print("history_obj NON contiene l'attributo 'history' o è None.")
else:
    print(f"history_obj contiene l'attributo 'history': {'history' in history_obj.__dict__}")
    print(f"Contenuto di history_obj.history: {history_obj.history}")

print("\nTest TabNet completato.")
Run the script in a PowerShell terminal: python test_tabnet.py

**Expected behavior**
The TabNetRegressor.fit() method should return a History object containing training and validation metrics, as described in the pytorch-tabnet documentation.

**Screenshots**
n/a

**Other relevant information:**
Operating System: [Il tuo sistema operativo e versione esatta, es. Windows 11 Pro, Build 22631.3798]
Python Version: Python 3.10.11
pytorch-tabnet Version: 4.1.0

Pip Freeze Output:
colorama==0.4.6
joblib==1.5.1
numpy==1.24.4
pandas==1.5.3
python-dateutil==2.9.0.post0
pytorch-tabnet==4.1.0
pytz==2025.2
scikit-learn==1.3.2
scipy==1.11.4
six==1.17.0
threadpoolctl==3.6.0
torch==1.13.1
tqdm==4.67.1
typing_extensions==4.14.1

**Additional context**
We have performed extensive troubleshooting, including:

Ensured C++ build tools for Visual Studio are installed. (Outcome: Confirmed, not the issue).

Addressed file locking issues by restarting the system. (Outcome: Resolved file locks, not the root cause).

Repeatedly recreated the virtual environment from scratch. (Outcome: Confirmed clean environment each time).

Used different Python versions:

Initially tried with Python 3.13 (global and in venv). (Outcome: ModuleNotFoundError for pandas or pytorch_tabnet, history_obj still None).

Switched to a fresh Python 3.10.11 installation for the virtual environment. (Outcome: Resolved Python 3.13 compatibility concerns, but history_obj still None).

Forced reinstallation of pytorch-tabnet and its dependencies multiple times. (Outcome: Packages installed correctly, but problem persisted).

Explicitly installed pandas when ModuleNotFoundError occurred. (Outcome: pandas issue resolved, but history_obj still None).

Manually uninstalled and then installed specific, older, stable versions of core dependencies (torch, numpy, pandas, scikit-learn, scipy) to ensure compatibility with pytorch-tabnet 4.1.0. (Outcome: Dependencies are now on known stable versions, but history_obj still None).

Verified active virtual environment: Confirmed (.venv) in PowerShell prompt at each step.

Checked pip freeze output multiple times to ensure correct package versions are installed (most recently confirming pytorch-tabnet's presence).

**Comment by lucalazzaro** [comment]
Hi @Optimox and @eduardocarvp , could you please support me with this or tell me if there's any missing information you might need? Thank you!

**Comment by Optimox** [comment]
This is not how things work. The fit method returns None, you want the history of your TabNet model. Please have a look at the demo notebooks as they show how to access the history.

```
clf = TabNetRegressor(**your_params_here)

clf.fit(your_data_here)

your_history = clf.history
```
